Mathlib/Analysis/Calculus/FDeriv/Comp.lean

```diff
@@ -0,0 +1,265 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Avigad, Sébastien Gouëzel, Yury Kudryashov
+
+! This file was ported from Lean 3 source module analysis.calculus.fderiv.comp
+! leanprover-community/mathlib commit e3fb84046afd187b710170887195d50bada934ee
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Analysis.Calculus.FDeriv.Basic
+
+/-!
+# The derivative of a composition (chain rule)
+
+For detailed documentation of the Fréchet derivative,
+see the module docstring of `Analysis/Calculus/FDeriv/Basic.lean`.
+
+This file contains the usual formulas (and existence assertions) for the derivative of
+composition of functions (the chain rule).
+-/
+
+
+open Filter Asymptotics ContinuousLinearMap Set Metric
+
+open Topology Classical NNReal Filter Asymptotics ENNReal
+
+noncomputable section
+
+section
+
+variable {𝕜 : Type _} [NontriviallyNormedField 𝕜]
+
+variable {E : Type _} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+
+variable {F : Type _} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+
+variable {G : Type _} [NormedAddCommGroup G] [NormedSpace 𝕜 G]
+
+variable {G' : Type _} [NormedAddCommGroup G'] [NormedSpace 𝕜 G']
+
+variable {f f₀ f₁ g : E → F}
+
+variable {f' f₀' f₁' g' : E →L[𝕜] F}
+
+variable (e : E →L[𝕜] F)
+
+variable {x : E}
+
+variable {s t : Set E}
+
+variable {L L₁ L₂ : Filter E}
+
+section Composition
+
+/-!
+### Derivative of the composition of two functions
+
+For composition lemmas, we put `x` explicit to help the elaborator, as otherwise Lean tends to
+get confused since there are too many possibilities for composition. -/
+
+
+variable (x)
+
+theorem HasFDerivAtFilter.comp {g : F → G} {g' : F →L[𝕜] G} {L' : Filter F}
+    (hg : HasFDerivAtFilter g g' (f x) L') (hf : HasFDerivAtFilter f f' x L) (hL : Tendsto f L L') :
+    HasFDerivAtFilter (g ∘ f) (g'.comp f') x L := by
+  let eq₁ := (g'.isBigO_comp _ _).trans_isLittleO hf
+  let eq₂ := (hg.comp_tendsto hL).trans_isBigO hf.isBigO_sub
+  refine' eq₂.triangle (eq₁.congr_left fun x' => _)
+  simp
+#align has_fderiv_at_filter.comp HasFDerivAtFilter.comp
+
+/- A readable version of the previous theorem, a general form of the chain rule. -/
+example {g : F → G} {g' : F →L[𝕜] G} (hg : HasFDerivAtFilter g g' (f x) (L.map f))
+    (hf : HasFDerivAtFilter f f' x L) : HasFDerivAtFilter (g ∘ f) (g'.comp f') x L := by
+  unfold HasFDerivAtFilter at hg
+  have :=
+    calc
+      (fun x' => g (f x') - g (f x) - g' (f x' - f x)) =o[L] fun x' => f x' - f x :=
+        hg.comp_tendsto le_rfl
+      _ =O[L] fun x' => x' - x := hf.isBigO_sub
+  refine' this.triangle _
+  calc
+    (fun x' : E => g' (f x' - f x) - g'.comp f' (x' - x)) =ᶠ[L] fun x' =>
+        g' (f x' - f x - f' (x' - x)) :=
+      eventually_of_forall fun x' => by simp
+    _ =O[L] fun x' => f x' - f x - f' (x' - x) := (g'.isBigO_comp _ _)
+    _ =o[L] fun x' => x' - x := hf
```

Comment:
```suggestion
/- porting note: todo: restore the example
Compile fails because `calc` fails to generate a `Trans` instance
example {g : F → G} {g' : F →L[𝕜] G} (hg : HasFDerivAtFilter g g' (f x) (L.map f))
    (hf : HasFDerivAtFilter f f' x L) : HasFDerivAtFilter (g ∘ f) (g'.comp f') x L := by
  unfold HasFDerivAtFilter at hg
  have :=
    calc
      (fun x' => g (f x') - g (f x) - g' (f x' - f x)) =o[L] fun x' => f x' - f x :=
        hg.comp_tendsto le_rfl
      _ =O[L] fun x' => x' - x := hf.isBigO_sub
  refine' this.triangle _
  calc
    (fun x' : E => g' (f x' - f x) - g'.comp f' (x' - x)) =ᶠ[L] fun x' =>
        g' (f x' - f x - f' (x' - x)) :=
      eventually_of_forall fun x' => by simp
    _ =O[L] fun x' => f x' - f x - f' (x' - x) := (g'.isBigO_comp _ _)
    _ =o[L] fun x' => x' - x := hf
-/
```